Add contributing guidelines and update README (#278)
mautic/developer-documentation-new: .github/CONTRIBUTING.md

# Contributing to Mautic developer documentation

Contributions are always welcome, no matter how large or small, or at whatever skill level you are. Before contributing, please read the [Code of Conduct](https://mautic.org/code-of-conduct/) and follow the directions in this guide.

---

## Table of contents

- [Communication expectation](#communication-expectation)
- [Issues](#issues)
- [Pull requests (PRs)](#pull-requests-prs)
  - [Before submitting a PR](#before-submitting-a-pr)
  - [Submitting a PR](#submitting-a-pr)
  - [After submitting a PR](#after-submitting-a-pr)
- [Contributing workflow](#contributing-workflow)
  - [Forking the repository](#forking-the-repository)
  - [Clone the repository](#clone-the-repository)
  - [Create a new branch](#create-a-new-branch)
    - [Ensure the correct base branch](#ensure-the-correct-base-branch)
    - [Ways to create a new branch](#ways-to-create-a-new-branch)
  - [Push changes to the remote repository](#push-changes-to-the-remote-repository)
  - [Create a PR](#create-a-pr)
    - [Git cherry-pick](#git-cherry-pick)
- [Getting started](#getting-started)
  - [1. On GitHub](#1-on-github)

>   - [2. GitHub Codespaces](#2-github-codespaces)

Reviewer: These need adding to the accept.txt in the singular and plural.

>     - [Setting up a codespace](#setting-up-a-codespace)

Author: It seems adding "codespace" (all lowercase) to the accept.txt doesn't fix the Vale error here.

Author: TIL that we can [ignore Vale in Markdown & HTML](https://docs.crossplane.io/contribute/vale/#ignore-all-rules). Should we apply this?

Reviewer: Yeah maybe. if it's an error it's going to keep failing. I think we'd need to exclude the whole ToC once we're happy everything else is fine, and if we can, add a comment explaining why we're turning off Vale.

  - [Create a new link](#create-a-new-link)
  - [Check broken links](#check-broken-links)
- [Working with Vale](#working-with-vale)
- [Adding a code sample](#adding-a-code-sample)
- [Updating contents and UI images](#updating-contents-and-ui-images)
- [Credit](#credit)

---

## Communication expectation

1. Always leave a detailed description in the pull request (PR). Leave nothing ambiguous for the reviewers.
2. Provide screenshots for visual changes.
3. Always review your code first. Be sure to run the project locally and test it thoroughly before requesting a review.
4. Communicate in the GitHub repository first before Slack. Whether it's in the issue or the PR, keeping the lines of communication open and visible to everyone on the team helps everyone around you.

## Issues

- When you contribute to the project for the first time, please consider checking the [good first issue](https://github.com/mautic/developer-documentation-new/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22good%20first%20issue%22) or [help wanted](https://github.com/mautic/developer-documentation-new/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22help%20wanted%22) labels.

> - If you wish to work on an open issue, please comment on the issue, and a maintainer will assign it to you.

Reviewer: ```suggestion
- If you wish to work on an open issue, please comment on the issue, and wait to be assigned by a maintainer.
```

  If an issue isn't assigned, it's assumed to be available for anyone to work on. So, ensure that you're assigned to an issue **before** beginning work to avoid conflicts.

>   **Note:** Please don't ask maintainers to assign you to another issue before you have finished working on yours and created a PR.

Reviewer: ```suggestion
  **Note:** please don't ask maintainers to assign you to another issue before you have finished working on yours and created a PR.
```

Could we use an admonition here?

3. Keep your branch up to date while waiting for review.
4. Respond and address the reviewer's feedback. Please don't request a review until you've addressed all feedback.

## Contributing workflow

In this section, you can find the contributing workflow and best practices for contributing to this project.

### Forking the repository

Forking this repository is the first step you need to take before proceeding. Forking means making a copy of a repository to your GitHub account.

To fork this repository:

1. In the main page of [mautic/developer-documentation-new](https://github.com/mautic/developer-documentation-new), click the 'Fork' button at the top.

   ![Fork button on GitHub](../assets/images/fork_button_github.png)

2. Choose your username in the 'Owner *' dropdown menu. **Don't select an organization here. Always choose your personal account**; otherwise, maintainers can't collaborate or fix issues in your PR.

   ![Choose fork owner on GitHub](../assets/images/choose_fork_owner_github.png)

> 3. Uncheck the option of 'Copy the DEFAULT-BRANCH-NAME branch only' because later on, you want to be able to clone multiple branches.

Reviewer: ```suggestion
3. Remove the check in option of 'Copy the DEFAULT-BRANCH-NAME branch only' because later on, you want to be able to clone multiple branches.
```

4. Click the green 'Create fork' button at the bottom.

>    ![An uncheck state option to copy only default branch and a create fork button on GitHub](../assets/images/uncheck_option_and_create_fork_button_github.png)

Reviewer: ```suggestion
   ![A check box with the removed check to copy only default branch and a create fork button on GitHub](../assets/images/uncheck_option_and_create_fork_button_github.png)
```

### Clone the repository

After you forked the repository, you need to clone it. Cloning means copying a repository to your local environment. In this case, you want to clone your forked repository.

> [!IMPORTANT]
>
> The Mautic Developer Documentation contains multiple branches that represent specific versions of Mautic. You should clone each branch into its own dedicated folder and make your changes within the appropriate folder.
>
> So, for example, when you need to make documentation changes for Mautic version 5, clone the branch `5.x` and save it in a folder. You can name the folder anything you want, but it's best to reflect the branch name. For example, use a name like `dev-docs-5`.

Follow the steps below to clone your forked repository:

1. Click your avatar on the top right.
2. Click 'Repositories'.

   ![Repositories option from a dropdown menu on GitHub](../assets/images/repositories_option_github.png)

3. Open your forked user-documentation repository. The URL should have your username. For example: `https://github.com/YOUR-GITHUB-USERNAME/developer-documentation-new`.
4. Click the green 'Code' button on top.

> 5. Select 'HTTPS' and copy the URL.

Reviewer: Poss needs adding to accept.txt.

   ![Highlight of code button, copy symbol, and HTTPS tab on GitHub](../assets/images/code_button_https_tab_github.png)

6. In your terminal, go to your local directory where you want to save this project.
7. Run the `git clone` command specifying the branch and folder name, and hit Enter:

   ```bash
   git clone --branch BRANCH-NAME https://github.com/YOUR-GITHUB-USERNAME/developer-documentation-new FOLDER-NAME
   ```

   Here's an example:

   ```bash
   git clone --branch 5.x https://github.com/YOUR-GITHUB-USERNAME/developer-documentation-new dev-docs-5
   ```

### Create a new branch

> Before making changes, ensure that you create a new branch and work on it. You don't want to directly work on the default — such as `main` — or any other base branch, because you won't be able to work on lots of things at the same time. If you make all those changes on one branch, you can't separate them and merge only one change at a time.

Reviewer: ```suggestion
Before making changes, ensure that you create a new branch and work on it. You don't want to directly work on the default - such as `main` - or any other base branch, because you won't be able to work on lots of things at the same time. If you make all those changes on one branch, you can't separate them and merge only one change at a time.
```

Making changes directly on GitHub is suitable for minor changes, such as fixing a typo. For bigger and more complex changes, please use GitHub Codespaces or work locally.

To work directly on GitHub, follow the steps below:

1. Click the 'Edit on GitHub' button on the top right of the page where you noticed the mistake. It takes you to the correct resource on GitHub.

   ![Screenshot of community handbook with a red box highlighting the Edit on GitHub button](../assets/images/edit_on_github.png)

2. Click the edit button — which resembles a pencil — and make the necessary changes.

   ![Screenshot of community handbook with a red box highlighting the Edit on GitHub button](../assets/images/edit_button_github.png)

3. Follow the instructions to commit the changes.
4. Select to commit to a new branch. Call the branch something relative to what you are updating.
5. [Create a PR](#create-a-pr). Read the "[Submitting a PR](#submitting-a-pr)" section about all info that you need to include in your PR.

### 2. GitHub Codespaces

Using GitHub Codespaces lets you quickly spin up the project in the cloud. For a smooth process, use the Chrome or Firefox browser to work with Codespaces.

> <details>

Reviewer: Is there a reason for dropping into HTML instead of Markdown here?

Author: Please see [my comment](https://github.com/mautic/developer-documentation-new/pull/278#issuecomment-3611286482) for the complete explanation. 🙂

Reviewer: Fabulous, thanks for explaining so clearly! :)

6. Wait until the codespace finishes building. Once it's finished, the build prompt closes, the README preview opens — you can close this preview, and the `postCreateCommand` runs. Please wait until it finishes doing its job.

   ![Screenshot of postCreateCommand running in terminal](../assets/images/postcreatecommand_on_terminal.png)
7. [Create a new branch](#create-a-new-branch) to work on your changes.

   **Info:** Once you create a new branch, it automatically switches to your new branch. If you haven't seen the branch changes in your terminal, run `git status`, and you should see your branch name.

8. Go to the `docs/` directory:

   ```bash
   cd docs
   ```

9. Find the folder and file that you need to work on.
10. Work on your changes and use the [live preview](#live-preview-on-codespace) to view and test your changes in real-time.
11. Ensure that the changes you made follow Mautic's style guide by running the Vale lint. Please read the "[Working with Vale](#working-with-vale)" section to use Vale.

#### Live preview on codespace

1. Ensure that you're in the `docs/` directory.
2. Run `make html`. It generates the `build` folder.

>    **Tip:** If you get `make: *** No rule to make target 'html'.  Stop.` error message after running the `make html` command, you should check if you're in the correct directory. You must be in the `docs/` directory to run this command.

Reviewer: Let's use the Tip admonition here.

Author: Same here. Please see [my comment](https://github.com/mautic/developer-documentation-new/pull/278#issuecomment-3611286482) for the complete explanation. 🙂

  <p>If the above steps fail:</p>
  <ol>
    <li>Close VS Code and the live preview browsers.</li>
    <li>Go to <a href="https://github.com/codespaces" target="_blank">https://github.com/codespaces</a>.</li>
    <li>At the bottom, you should see a list of your codespaces.</li>
    <li>Click the three dots icon on the right of your project's codespace.</li>
    <li>Click 'Stop codespace'.</li>
    <li>Re-open the codespace by clicking its name.</li>
    <li>Follow the steps in the <a href="#live-preview-on-codespace">Live Preview on Codespace</a> section.</li>
  </ol>
</details>

<br />

> [!TIP]
>
> - Always refresh the page to view the new changes you have applied.
> - All commands only work within the `docs/`directory. If you're unable to run a command, verify that you're in the correct directory.
> - Read the "[Troubleshooting live preview](#troubleshooting-live-preview)" section if you encounter any issues with the live preview in the codespace.

<br />

If you're ready to push your changes to the remote repository and create a PR, please read the "[Push changes to the remote repository](#push-changes-to-the-remote-repository)" and "[Create a PR](#create-a-pr)" sections.

> ### 3. Local development

Reviewer: I feel like for the dev docs we should lead with working locally, perhaps? Up to you though! :)

Author: If it's up to me, I'd like to encourage folks to work locally for all docs! 😄 

I mean, the setup process might be quite complicated for folks who are not familiar with working locally, but once it's up and running, it's _much easier_ to work locally than on codepsace. Also, we have a video now that explaining how to set up local env.

Should we remove the codespace section from here then? And probably link the video as well to the local dev.

Reviewer: I would probs put the working locally section first then have the codespaces as an 'if you can't work locally, use this' option.

Author: Gotcha. Let me figure that out. 👍🏼

   ```rst
   .. vale off

   * All PRs are made against the ``c.x`` branch in the first instance, for instance, ``5.x``.
   * If the PR should be merged in an earlier release than the next major release of Mautic, duplicate the PR against the relevant ``a.b`` branch for bug fixes - for example, ``5.0`` - or ``a.x`` branch for features and enhancements - for example, ``5.x``.
   * Backwards compatibility breaking changes can only be released in a major version, so they should only ever be made against the ``c.x`` branch, such as, ``5.x``.

   .. vale on
   ```

<br />

> [!IMPORTANT]
>
> - Ensure that you wrap the sentences that you'd like Vale to skip with both `.. vale off` and `.. vale on` statements, in order. Failing to do so results in Vale lint skipping the rest of the contents.
> - Don't add statements to skip lint, unless necessary. If you're uncertain, it's best not to wrap them in the statements and let the team review and provide suggestions.

## Adding a code sample

Code samples get downloaded from GitHub to ensure that they're always up to date. If you want to add a new code sample, follow these steps:

> 1. Create a `.py` file in the `docs/code_samples/` folder. For example: `helloworld_entity_world.py`.

Reviewer: I don't think you need to do this. You can just use `make code-sample` in the terminal, according to the [makefile](https://github.com/mautic/developer-documentation-new/blob/5.x/docs/Makefile#L39), just like the `make link` command.

Author: I only moved this section from README and make the process clearer because I never worked with code samples. Let me try the ` make code-sample` out and rewrite this section. 👍🏼

4. Click the green 'Create fork' button at the bottom.

   ![A check box with the removed check to copy only default branch and a create fork button on GitHub](../assets/images/uncheck_option_and_create_fork_button_github.png)

### Clone the repository

After you forked the repository, you need to clone it. Cloning means copying a repository to your local environment. In this case, you want to clone your forked repository.

> [!IMPORTANT]
>
> The Mautic Developer Documentation contains multiple branches that represent specific versions of Mautic. You should clone each branch into its own dedicated folder and make your changes within the appropriate folder.
>
> So, for example, when you need to make documentation changes for Mautic version 5, clone the branch `5.x` and save it in a folder. You can name the folder anything you want, but it's best to reflect the branch name. For example, use a name like `dev-docs-5`.

Follow the steps below to clone your forked repository:

1. Click your avatar on the top right.
2. Click 'Repositories'.

   ![Repositories option from a dropdown menu on GitHub](../assets/images/repositories_option_github.png)

> 3. Open your forked user-documentation repository. The URL should have your username. For example: `https://github.com/YOUR-GITHUB-USERNAME/developer-documentation-new`.

Author: ```suggestion
3. Open your forked developer-documentation-new repository. The URL should have your username. For example: `https://github.com/YOUR-GITHUB-USERNAME/developer-documentation-new`.
```

- If you want to work on an open issue, please comment on it so a maintainer can assign it to you.

  If an issue isn't assigned, it's assumed to be available for anyone to work on. So, ensure that you're assigned to an issue **before** beginning work to avoid conflicts.

- Don’t ask maintainers to assign you to another issue before you finish working on your current one and create a PR. Also, avoid requesting assignment to an issue that already has someone assigned. However, if the assignee hasn’t addressed the issue for a while and you’re interested in working on it, leave a comment to ask about its status and progress.

- Did you spot a typo, missing instructions, or have an idea for enhancing the Mautic developer documentation? You can [create an issue](https://github.com/mautic/developer-documentation-new/issues/new/choose) to address it.

  However, the Education Team needs to triage the issue before you can work on it. If you wish to work on the issue you submitted, please inform and tag the `@mautic/education-team-leaders` in the comment.

## Pull requests

> Pull requests - PRs - are always welcome. However, before working on changes, you must ensure that **you are assigned** to an existing issue and **link your work to the issue in your PR**.

Reviewer: ```suggestion
Pull requests - PRs - are always welcome. However, before working on changes, you must ensure that a maintainer **assigns you** to an existing issue, and always **link your work to the issue in your PR**.
```

   - **A related issue**. [Link the issue number](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) that you worked on and add a keyword of 'Closes', 'Fixes', or 'Resolves' in front of it. For example, `Closes #123`, `Fixes #234`, etc. You can find the issue number right next to the issue's title. When you link the issue number, the issue automatically closes once a maintainer merges your PR.

3. Provide screenshots for visual changes if necessary.

> [!TIP]
>
> Refer to [PR #369](https://github.com/mautic/mautic-community-handbook/pull/369) in the Mautic Community Handbook for an example of a well-documented PR.

### After submitting a PR

1. Ensure that all checks pass. If the linting build or prose fails, debug and fix it until all passes. If you have questions or need help, feel free to tag the `@mautic/education-team-leaders` in the comment.
2. Please don't DM maintainers on Slack to review or ask feedback and questions about your PR.

>    If you'd like feedback or ask questions about your PR, tag `@mautic/education-team-leaders` in the comment of your PR or use the `#t-education` channel on Slack. That way, not only maintainers, but the community can help you get unstuck. The team always receives a notification whenever an incoming PR is submitted. If you haven't received a review within a week, you can tag them in the PR comments to ask for an estimated review time.

Reviewer: ```suggestion
   If you'd like feedback or ask questions about your PR, tag `@mautic/education-team-leaders` in the comment of your PR or use the `#t-education` channel on Slack. That way, not only maintainers, but the community can help you get unstuck. The team always receives a notification on new PR creation. If you haven't received a review within a week, you can tag them in the PR comments to ask for an estimated review time.
```
